endpoint_route_handler/registry.py

```diff
@@ -0,0 +1,342 @@
+                END;
+                $$ LANGUAGE plpgsql;
+            """
+            )
+            cr.execute(
+                """
+                CREATE TRIGGER trigger_endpoint_route_set_timestamp
+                BEFORE UPDATE ON endpoint_route
+                FOR EACH ROW
+                EXECUTE PROCEDURE endpoint_route_set_timestamp();
+            """
+            )
+
+    def __init__(self, cr):
+        self.cr = cr
+
+    def get_rules(self, keys=None, where=None):
+        for row in self._get_rules(keys=keys, where=where):
+            yield EndpointRule.from_row(self.cr.dbname, row)
+
+    def _get_rules(self, keys=None, where=None, one=False):
+        query = "SELECT * FROM endpoint_route"
+        pargs = ()
+        if keys and not where:
+            query += " WHERE key IN %s"
+            pargs = (tuple(keys),)
+        elif where:
+            query += " " + where
+        self.cr.execute(query, pargs)
+        return self.cr.fetchone() if one else self.cr.fetchall()
+
+    def _get_rule(self, key):
+        row = self._get_rules(keys=(key,), one=True)
+        if row:
+            return EndpointRule.from_row(self.cr.dbname, row)
+
+    def _lock_rows(self, keys):
+        sql = "SELECT id FROM endpoint_route WHERE key IN %s FOR UPDATE"
+        self.cr.execute(sql, (tuple(keys),), log_exceptions=False)
+
+    def _update(self, rows_mapping):
+        self._lock_rows(tuple(rows_mapping.keys()))
+        return query_multi_update(
+            self.cr,
+            self._table,
+            tuple(rows_mapping.values()),
+            EndpointRule._ordered_columns(),
+        )
+
+    def _create(self, rows_mapping):
+        return query_insert(self.cr, self._table, list(rows_mapping.values()))
+
+    def get_rules_by_group(self, group):
+        rules = self.get_rules(where=f"WHERE route_group='{group}'")
+        return rules
+
+    def update_rules(self, rules, init=False):
+        """Add or update rules.
+
+        :param rule: list of instances of EndpointRule
+        :param force: replace rules forcedly
+        :param init: given when adding rules for the first time
+        """
+        keys = [x.key for x in rules]
+        existing = {x.key: x for x in self.get_rules(keys=keys)}
+        to_create = {}
+        to_update = {}
+        for rule in rules:
+            if rule.key in existing:
+                to_update[rule.key] = rule.to_row()
+            else:
+                to_create[rule.key] = rule.to_row()
+        res = False
+        if to_create:
+            self._create(to_create)
+            res = True
+        if to_update:
+            self._update(to_update)
+            res = True
+        return res
+
+    def drop_rules(self, keys):
+        self.cr.execute("DELETE FROM endpoint_route WHERE key IN %s", (tuple(keys),))
+        return True
+
+    def make_rule(self, *a, **kw):
+        return EndpointRule(self.cr.dbname, *a, **kw)
+
+    def last_update(self):
```

Comment:
should be replaced by `def last_version(self)`

```suggestion
    def last_version(self):
          self.cr.execute("SELECT last_value FROM endpoint_version")
          return self.cr.fetchone()[0]
```

Reply:
I introduced a sequence on v14 here https://github.com/OCA/web-api/pull/3/commits/3ae6ceb6ef6ba83682b5d9037792cf5d3fba178d

I'm testing. I'll fwd port soon.